### DISCUSSION TITLE
Thread 'Thread-8 (_listening)': missing ScriptRunContext Deepgram STT

### DISCUSSION THREAD

**Discussion by Emaa719** [deepgram/community#709] [General help] [answered]
### Which Deepgram product feature are you using?

Deepgram API - STT Streaming

### Details

Hello, I am working on a voice chatbot using Deepgram for streaming speech-to-text (STT), an OpenAI model for natural language understanding, and ElevenLabs for text-to-speech (TTS). The code in Visual Studio Code (VSCode) is working well, but I'm encountering issues when deploying the chatbot.

During deployment, the chat function and the TTS generation are working without errors. However, the real-time transcription function using Deepgram for recording and transcribing text is encountering errors.

I am deploying the chatbot using Streamlit, and I have also tried using Gradio without success.

this is the error i am getting '

Press Enter to stop recording...

2024-04-17 15:21:47.948 Thread 'Thread-8 (_listening)': missing ScriptRunContext
2024-04-17 15:21:47.951 Thread 'Thread-8 (_listening)': missing ScriptRunContext
2024-04-17 15:21:47.953 Thread 'Thread-8 (_listening)': missing ScriptRunContext
2024-04-17 15:21:47.954 Thread 'Thread-8 (_listening)': missing ScriptRunContext
speaker:'

### If possible, please attach your code or paste it into the text box.

from elevenlabs.client import ElevenLabs
from elevenlabs import stream
from dotenv import load_dotenv
from langchain_openai import ChatOpenAI
#import logging, verboselogs
from langchain.memory import ConversationBufferMemory

from time import sleep
import os
from deepgram import (
    DeepgramClient,
    Deepgram,
    DeepgramClientOptions,
    LiveTranscriptionEvents,
    LiveOptions,
    Microphone,
)
import requests
import re, elevenlabs
from colorama import Fore, Style, init
from pydub import AudioSegment
from pydub.playback import play
from langchain.prompts import ChatPromptTemplate

import openai

conversation1 = []  
def main():
    try:

        deepgram: DeepgramClient = DeepgramClient(DEEPGRAM_API_KEY)

        dg_connection = deepgram.listen.live.v("1")

        full_transcript = ""

        def on_open(self, open, **kwargs):
            print(f"\n\n{open}\n\n")

        def on_message(self, result, **kwargs):
            sentence = result.channel.alternatives[0].transcript
            if len(sentence) !=0:
                nonlocal full_transcript  # Indiquer que nous modifions la variable extérieure
                full_transcript += sentence + " "
            '''if len(sentence) == 0:
                return'''
            print(f"speaker: {sentence}")
            

        def on_metadata(self, metadata, **kwargs):
            print(f"\n\n{metadata}\n\n")

        def on_speech_started(self, speech_started, **kwargs):
            print(f"\n\n{speech_started}\n\n")

        def on_utterance_end(self, utterance_end, **kwargs):
            print(f"\n\n{utterance_end}\n\n")

        def on_close(self, close, **kwargs):
            print(f"\n\n{close}\n\n")

        def on_error(self, error, **kwargs):
            print(f"\n\n{error}\n\n")

        def on_unhandled(self, unhandled, **kwargs):
            print(f"\n\n{unhandled}\n\n")

        dg_connection.on(LiveTranscriptionEvents.Open, on_open)
        dg_connection.on(LiveTranscriptionEvents.Transcript, on_message)
        dg_connection.on(LiveTranscriptionEvents.Metadata, on_metadata)
        #dg_connection.on(LiveTranscriptionEvents.SpeechStarted, on_speech_started)
        dg_connection.on(LiveTranscriptionEvents.UtteranceEnd, on_utterance_end)
        dg_connection.on(LiveTranscriptionEvents.Close, on_close)
        dg_connection.on(LiveTranscriptionEvents.Error, on_error)
        #dg_connection.on(LiveTranscriptionEvents.Unhandled, on_unhandled)

        options: LiveOptions = LiveOptions(
            model="nova-2-general",
            punctuate=True,
            language="fr",
            encoding="linear16",
            channels=1,
            sample_rate=16000,
            # To get UtteranceEnd, the following must be set:
            interim_results=True,
            utterance_end_ms="1000",
        )

        print("\n\nPress Enter to stop recording...\n\n")
        if dg_connection.start(options) is False:
            print("Failed to connect to Deepgram")
            return

        # Open a microphone stream on the default input device
        microphone = Microphone(dg_connection.send)

        # start microphone
        microphone.start()

        # wait until finished
        input("")

        # Wait for the microphone to close
        microphone.finish()

        # Indicate that we've finished
        dg_connection.finish()
        print("Full transcript:", full_transcript)
        return full_transcript
        print("Finished")
        # sleep(30)  # wait 30 seconds to see if therae is any additional socket activity
        # print("Really done!")
        

    except Exception as e:
        print(f"Could not open socket: {e}")
        return
    
prompt = ChatPromptTemplate.from_messages([
    ("system", ' '),
    ("user", "{input}")
])
chat_memory = ConversationBufferMemory()
conversation = []
def run_chat_agent(user_input, conversation):
    conversation.append({"role": "user", "content": user_input})
    llm = ChatOpenAI(model="gpt-3.5-turbo", temperature=0.9)
    chain = prompt | llm
    result = chain.invoke({"input": user_input})
    conversation.append({"role": "assistant", "content": result})
    return result.content
# Streamlit interface

import streamlit as st    

if st.button('Cliquez ici'):
    t= main()
    response = run_chat_agent(t, conversation)

**Comment by team-deepgram** [comment]
Thanks for asking your question about Deepgram! If you didn't already include it in your post, please be sure to add as much detail as possible so we can assist you efficiently, such as:
- The `request_id` if you have a question about your requests or transcription responses.
- The features you used or the full `api.deepgram.com` URL you sent your request to, including parameters.
- Any code snippets you can share.

**Comment by davidvonthenen** [comment, marked as answer]
hi @ImaneMottaqi 

Streamlit is a hosted service which won't work with the Microphone example as is. I don't think you would be able to use the local microphone on the Streamlit host where it is running.

If you want an example of running something similar but on your local laptop:
https://github.com/dvonthenen/proposals/blob/main/2024_SCALE21/VoiceActivatedAICollaborators/DEMO-Voice-AI-Assistant/Deepgram/deepgram-openai.py
